jax/_src/interpreters/pxla.py

```diff
@@ -2127,7 +2131,8 @@ def _to_logical_sharding(
   elif isinstance(aval, core.AbstractToken):
     return None
   else:
-    raise TypeError(aval)
+    assert isinstance(sharding, sharding_impls.XLACompatibleSharding)
```

Comment:
I would suggest creating another elif for dynamic aval instead of returning sharding by default.

Reply:
Done.